feat: Hierarchical Model Importer

noa7 opened the pull request:
This PR adds an “Split hierarchy” workflow for model import: when enabled, Stride imports one Model per mesh (named after the mesh) and automatically generates a Prefab that mirrors the FBX node hierarchy. When disabled, the import produces a single combined Model (no “(All)” suffix).

What’s in this PR-

Split hierarchy checkbox in Add asset → 3D model dialog.

Remembers last value per project.

Behavior (per mode)

Split hierarchy = ON

Imports N Model assets (one per mesh) named:
"<ModelName>-<MeshName>" (sanitized; auto-uniquified with -1, -2, …).

Creates a Prefab named "<ModelName> Prefab" that replicates the FBX hierarchy (one Entity per node, correct parent/child relationships).

Each Entity that hosts a mesh gets a ModelComponent referencing the corresponding per-mesh Model.


Each per-mesh Model keeps only the material(s) it actually uses\

Split hierarchy = OFF

Imports a single Model named exactly "<ModelName>"

Eideren committed:
Nullable UrlReference for scenes, empty is not allowed by the serializer

Eideren committed:
Fix duplicate import warnings

Eideren committed:
Remove unnecessary nuget config

Eideren committed:
Remove unnecessary call to base implementation from empty virtual methods

Eideren committed:
Create one default simulation when there isn't any configuration setup

Eideren committed:
Remove demo projects from the main solution

Eideren committed:
awaitable simulation NextTick() and AfterTick()

Eideren committed:
Fix contact event on collidable removal

Eideren committed:
Add Contact Removal and Raycast removal tests

Eideren committed:
Add summary for "collisionMask" in ray tests + fix RaycastPenetrating naming scheme

Eideren committed:
Fix demo after naming scheme change

Eideren committed:
Align *Tick() naming scheme with ISimulationUpdate

Jklawreszuk committed:
feat: Make Stride.StorageTool as Gui Tool (#2479)

* feat: Make Stride.StorageTool as Gui Tool

* fixup! feat: Make Stride.StorageTool as Gui Tool

* fixup! feat: Make Stride.StorageTool as Gui Tool

* Update README.md

---------

Co-authored-by: Nicolas Musset <musset.nicolas@gmail.com>

kopffarben committed:
fix: Wrong Instanced call used for DrawInstanced in CommandList.Direct3D.cs (#2482)

CommandList.DrawInstanced(Buffer argumentsBuffer, int alignedByteOffsetForArgs = 0)
NativeDeviceContext was DrawIndexedInstancedIndirect but should be DrawInstancedIndirect

Eideren committed:
Fix manifold and improve character component handling

Eideren committed:
Character component: cut custom gravity, bepu will take care of it

Eideren committed:
Improve camera controls for demo character controller

Doprez committed:
Bepu Navigation fixes (#6)

* navigation component overhaul

* Removing redundant code

Doprez committed:
Bepu navigation clean up part 2 (#7)

* navigation component overhaul

* Removing redundant code

* move to GameSystem

* Updated DotRecast

* fix world transofrm move bug

* added build time property

* More settings!

* fix forward to be Z

* Eiderens a genius fixed move

* docs and extra options for tryfindpath

* omit string value

Jklawreszuk committed:
fix: Adjust Solution.cs for all platforms (#2472)

Jklawreszuk committed:
fix: Enable to build SkyBoxAsset using OpenGL (#2504)

Jklawreszuk committed:
fix: Resolve crashing issue for OpenGL UIPage (#2189) (#2500)

Eideren committed:
Revert "[Editor] Collection supports in abstract properties" as it is no longer necessary.

This reverts commit 3f710c690faa7600dc19daceff898200e1977e58.
This commit is part of another PR on Stride/master which will be merged at a later date.

Eideren committed:
Merge remote-tracking branch 'STRIDE/master' into bepu_integration

# Conflicts:
#	sources/Directory.Packages.props

Eideren committed:
Removed SimulationUpdateComponent in favor of ISimulationUpdate

Eideren committed:
Pass CollidableComponent instead of CollidableReference to IContactEventHandler - using CollidableReference in userland is unsafe

dloe committed:
Test case improvements and updates (#2393)

* TestHalf2 and TestHalf now can pass in suite

Half2 now has proper ToSting (before it would simply  output the type instead of the values with ToString). Converter seems to similarly match formats for other converts to a degree

* Added similar prints for Half3 and Half4, added comments on Converter

- Need verification on how the converters should be changed (if they even still need to be)
- All half tests now pass

* LifeTimeNoSimpleConstructor now passes

Test now passes but a bit confused as to what this  test is trying to do. Will need context with PR. Didnt have empty constructor therefore it was impossible for this to ever pass.

* RenderToWindow test now skips due to failure

Relates to a possible rework in GameTestBase and SendImage method

* Various Graphics Tests Using outdated images [DONT USE]

10 Windows Tests Updated
11 Windows Tests Updated

RunTestGame AMD test animation images are now uodated

RunTestGame seems to be using older images for animation tests

SpriteRender2DTests has outdated images

Updated SpriteRenderer2DTests.f1 and Tests files
(Test was failing for small pixel differences)

SpriteRenderer3DTests Updated Images

- Images were pixels off causing a failure

SpriteRotationTests updated images

- Images were pixels off causing a failure

SpriteTestGame updated images

- Images were pixels off causing a failure

TesselationTests updated images

- Images were pixels off causing a failure

SceneAmbientLight test has outdated images

Updated SceneAmbientLight image

* TEstGCHandleAlloc: Autolayout was set to fail when it does now in fact succeed

Unsure why this originally was set to false after reading documentation, might be missing some context. Can simply add a TODO above line 37 instead to give more context. The rest of the test seems to function as expected.

* Stride Particles Tests Updated Visual Tests

- Outdated images for various tests inside Particles.Tests.Windows

Stride.Graphics.Tests - Updated image for TestRenderToTexture Test

- Outdated images for AMD

* Updated TestGetFontInfo to now properly check if various Font Info is properly loaded

* Added skips for graphics related tests to avoid confusion and give additional info for why they fail info

* Updated template package names to now be properly loaded

For whatever reason, it was case sensitive and could not find the corresponding template packages. We not hit another error for these tests but it gets much farther now.

* Revert "Stride Particles Tests Updated Visual Tests"

This reverts commit 54e0570601569aa3180dfa5609c08aaf568c43be.

* Revert "Various Graphics Tests Using outdated images [DONT USE]"

This reverts commit dc984b593cea9e8b8752f5da80cfa7b3b89beb89.

* Added skips to GPU rendering related tests that will consecutively fail

For the time being, since these tests are designed for the teamcity agent and that isnt operational, these can be skipped to avoid large amount of failures.

Note: May need feedback on these, might be okay to not gave these skip, regardless of if they always fail.

* Skipped 2 more rendering related test cases, cleaned comments

* Revert "Added skips to GPU rendering related tests that will consecutively fail"

This reverts commit 110a9161ac086c391792e061139d95e04cfc9278.

* Revert "Added skips for graphics related tests to avoid confusion and give additional info for why they fail info"

This reverts commit d598082afe6dd4e9b53ac2102cdc87b8a1b8ece7.

* Removed changed tests back to original tests with more comments

Hopefully can give insight for those seeing issues while SharpFont issues are being looked at

* Revert "Updated template package names to now be properly loaded"

This reverts commit 51c887b0dc0f788ff50f25f1948975af5ae0e166.

* Removed Skips for rendering

* Removed LifetimeNoSimpleConstuctor test and unnecessary constructor

Eideren committed:
Deprecate DataPointer in favor of Span (#2421)

Eideren committed:
feat: TransformComponent.SetWorld, strict equality and fix misc. typos (#2503)

* feat: TransformComponent.SetWorld and fix misc. typos

* Skip unnecessary struct copy

Feralnex committed:
Fixed getting actual bounds and refresh rate of the output (monitor/display) from GraphicsOutput.CurrentDisplayMode when using Direct3D (#2494)

* Fixed getting actual bounds and refresh rate of the output (monitor/display) from GraphicsOutput.CurrentDisplayMode when using Direct3D

* Added exception logging when failed to create Direct3D device in GraphicsOutput.

* Return null if current format doesn't match the requested one

* CurrentDisplayMode initialization refactor

Basewq committed:
[Build] Update Silk.NET from 2.20.0 to 2.22.0 (#2521)

Co-authored-by: Basewq <Basewq@users.noreply.github.com>

Eideren committed:
Remove projects still in WIP state from the solution

Eideren committed:
Merge pull request #2131 from Eideren/bepu_integration

feat: Bepu physics integration

Eideren committed:
fix: Add in gizmos on assembly load (#2526)

Jklawreszuk committed:
refactor: FreeImage dependency update (#2531)

* Refactor freeimage files

* Remove unnecessary code

* Remove unused FreeImage wrapper

* fixup! Remove unused FreeImage wrapper

Jklawreszuk committed:
feat: Get LocalAppData path from Environment.SpecialFolder obj (#2511)

VaclavElias committed:
build: Update NuGet.* packages to version 6.12.1 (#2537)

Eideren committed:
fix: Bepu, ensure users require an explicit reference to bepu before being able to use related components and assets (#2534)

Eideren committed:
feat: ExecutionMode control for Flexible Processing (#2535)

Jklawreszuk committed:
feat: Port predicates.c to .NET (#2547)

* Remove unused code

* Fix warnings

* Port predicates.c to .NET

* Add license

* Exclude Stride.Native proj from Stride.Rendering

* fixup! Exclude Stride.Native proj from Stride.Rendering

TranquilAbyss committed:
feat: Made UIElement Key Events public (#2518)

VaclavElias committed:
docs: README.md update - Consistency updates (#2473)

* feat: Readme updates

* feat: Content updates

* feat: Content updated

* feat: Readme updated

* feat: Content updated

* feat: Content improvements

* fix: Some content updates

* feat: Content improvements

* feat: Content updated

* fix: Content updated

* feat: Content updated

* feat: if-building-fails added

* feat: Content updates

* fix: Note word removed

* Update README.md

Co-authored-by: Jakub Ławreszuk <31008367+Jklawreszuk@users.noreply.github.com>

* Update README.md

* Update CODE_OF_CONDUCT.md - Empty line added for better readability.

* Update README.md - Badge added temporarily

* docs: Updates made based on the feedback

---------

Co-authored-by: Jakub Ławreszuk <31008367+Jklawreszuk@users.noreply.github.com>

riibaisu committed:
Replace Xenko logo in MaterialModel.fbx to Stride logo (#1505) (#2546)

Model by CodingSea/MaterialModel@6ea5af27b4d211e699b8b6a45c112d442815d0e4

VaclavElias committed:
chore: Update ServiceWire package to version 5.6.0 (#2548)

- Updated the ServiceWire package version in Directory.Packages.props from 5.5.4 to 5.6.0.

xen2 committed:
[Assets] Fix Stride project upgrader (broken for version 4.0 => 4.2)

Eideren committed:
feat: Process required keyword like MemberRequiredAttribute (#2555)

* feat: Process required keyword like MemberRequiredAttribute

* Add tests

Eideren committed:
fix: very minor prop. grid presentation changes for bepu (#2556)

Jklawreszuk committed:
fix: Remove unused using directive in Stride.Audio (#2557)

Eideren committed:
fix: Bepu, correct minor XML issue (#2558)

Eideren committed:
fix: Remove nuget config workaround for demo project (#2559)

Eideren committed:
fix: CharacterComponent jumping inside trigger boxes (#2562)

Basewq committed:
[Android] Fix invalid configuration in Android solution file. (#2563)

Co-authored-by: Basewq <Basewq@users.noreply.github.com>

Jklawreszuk committed:
feat: LockFileEx/UnLockFileEx methods shouldn't be called explicitly (#2564)

TranquilAbyss committed:
fix: Removed Focused loss triggering button Release for keyboard and mouse (#2549)

azeno committed:
[Graphics] Fixes memory leak when creating D3D11 texture with shared options

Nicogo1705 committed:
Allow user to set thread count (#2571)

Eideren committed:
Revert "chore: Update ServiceWire package to version 5.6.0 (#2548)" (#2575)

This reverts commit d9df9dd0be332ef442024349a6586eee069d1b28.

ourabigdev committed:
feat: added outline support for TextBlock (#2582)

* added outline support for TextBlock

added the possibility to add outline to you text block OutlineColor to change the color and OutlineThickness to change Thickness

* Optimize TextBlock outline rendering using nested loops

- No array allocation (Vector2[]) - reduces garbage collection pressure
- More efficient memory usage - uses stack instead of heap

* Remove unrelated whitespace

* Use empty alpha instead of null comparison - color is a struct

---------

Co-authored-by: Eideren <contact@eideren.com>

Doprez committed:
feat: Add numerics type conversions (#2238)

Eideren committed:
docs: Bepu, missing property summaries (#2567)

Eideren committed:
fix: UrlReference triggering quantum change propagation, leading to exceptions with cyclic scene references (#2528)

Jklawreszuk committed:
docs: Update THIRD PARTY.md file (#2577)

Eideren committed:
fix: GridLogViewer empty when messages are appended before the element is on screen (#2525)

Jklawreszuk committed:
feat: Port NuGetResolver.UI to AvaloniaUI (#2512)

* feat: Port NuGetResolver.UI to AvaloniaUI

* fixup! feat: Port NuGetResolver.UI to AvaloniaUI

* fixup! feat: Port NuGetResolver.UI to AvaloniaUI

ClamEater14 committed:
fix: Check repeat key flag before toggling full screen (#2593)

Eideren committed:
fix: bepu, removal of collidables while running a contact event leaves contact tracking in an unexpected state (#2603)

* fix: Re-entry/removal of collidables in contact events

* Finish Flush implementation, add tests

* Update comments and name of SetFreshnessForCurrentActivityStatus

Eideren committed:
Core: ServiceRegistry, GetOrCreate<T> and Remove(instance) (#2599)

Eideren committed:
Core: Refactor service usage following the changes introduced in PR #2599 (#2600)

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2607 from stride3d/all-contributors/add-Nicogo1705

docs: add Nicogo1705 as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2608 from stride3d/all-contributors/add-ourabigdev

docs: add ourabigdev as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2610 from stride3d/all-contributors/add-kopffarben

docs: add kopffarben as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2612 from stride3d/all-contributors/add-Feralnex

docs: add Feralnex as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2613 from stride3d/all-contributors/add-TranquilAbyss

docs: add TranquilAbyss as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2614 from stride3d/all-contributors/add-levifmorais

docs: add levifmorais as a contributor for design

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2615 from stride3d/all-contributors/add-tymokvo

docs: add tymokvo as a contributor for doc

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2616 from stride3d/all-contributors/add-Arc-huangjingtong

docs: add Arc-huangjingtong as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2617 from stride3d/all-contributors/add-minktusk

docs: add minktusk as a contributor for code, and content

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2618 from stride3d/all-contributors/add-timcassell

docs: add timcassell as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2619 from stride3d/all-contributors/add-dloe

docs: add dloe as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2620 from stride3d/all-contributors/add-wrshield

docs: add wrshield as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2621 from stride3d/all-contributors/add-soorMSWE

docs: add soorMSWE as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2622 from stride3d/all-contributors/add-MechWarrior99

docs: add MechWarrior99 as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2623 from stride3d/all-contributors/add-kristian15959

docs: add kristian15959 as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2624 from stride3d/all-contributors/add-YerkoAndrei

docs: add YerkoAndrei as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2625 from stride3d/all-contributors/add-ComputerSmoke

docs: add ComputerSmoke as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2626 from stride3d/all-contributors/add-timconner

docs: add timconner as a contributor for code

VaclavElias committed:
Update .all-contributorsrc

VaclavElias committed:
Update README.md

VaclavElias committed:
Merge pull request #2627 from VaclavElias/all-contributors-fydar

docs: Contributor details updated

Jklawreszuk committed:
feat: Crossplatform font loading (#2530)

johang88 committed:
feat: Added support for per frame logical groups in RootEffectRenderFeature (#2597)

VaclavElias committed:
docs: Bepu, XML Comments improvements (#2589)

johang88 committed:
feat: Generalized per view resource binding (#2609)

Basewq committed:
[Build] Fix missing version update of Silk.NET to version 2.22.0 (#2632)

Co-authored-by: Basewq <Basewq@users.noreply.github.com>

Eideren committed:
Revert "feat: Crossplatform font loading (#2530)" (#2637)

This reverts commit a06300f5fc261b65ebcf4e9fe1180162462d9cab.

Basewq committed:
fix: Legacy DynamicNavigationMeshSystem bugs and memory leak (#2631)

* [Navigation] Fix detecting DynamicNavigationMeshSystem in NavigationProcessor and fix when DynamicNavigationMeshSystem is disabled

* [Navigation] Ensure native resources are properly disposed from Bullet & Recast

---------

Co-authored-by: Basewq <Basewq@users.noreply.github.com>

Basewq committed:
[Android] Fix registering broadcast receiver on Android 14+ (#2638)

Co-authored-by: Basewq <Basewq@users.noreply.github.com>

net2cn committed:
fix: TextBlock enters an infinite loop if the given available space is too small to fit even one character. (#2598)

Signed-off-by: net2cn <mcopener@gmail.com>

Eideren committed:
fix: Add assets cleanup target when calling dotnet clean, clarify ref out of sync issues (#2629)

Kryptos-FR committed:
refactor: Nullable and Modernization for Stride.Core. & Stride.Core.AssemblyProcessor (#2639)

Eideren committed:
fix: Rendering hang when creating object node (#2640)

Kryptos-FR committed:
refactor: Nullable and Modernization for Core projects (#2643)

* [Core.Mathematics] Modernize code

* [Core.Mathematics] Add tests to Stride.Tests.Simple

* [Core] Cleanup some code

* [Core.CompilerServices] Cleanup some code

* [Core.CompilerServices] Add tests to Stride.Tests.Simple

* [Core.IO] Modernize code

* [Core.MicroThreading] Modernize code

* [Core.Serialization] Modernize code

* [Core.AssemblyProcessor.Tests] Modernize code

* [Core.CompilerServices.Tests] Modernize code

* [Core.Mathematics.Tests] Modernize code

* [Core.Tests] Modernize code

* [Core.Translation] Modernize code

* [Core.Reflection] Modernize code

* Fixes build of Stride.VisualStudio.sln

* [Core.Yaml.Tests] Modernize code

* [Core.Design] Modernize code

* [Core.Design.Tests] Modernize code

* [Core.Mathematics] More annotations

* Fix build of projects that import code from Core.Design

Especially those targeting .NET Framework

Eideren committed:
feat: Add asset support for interface properties (#2606)

Kryptos-FR committed:
fix: Temporarily disable the resolver UI as there is a dependency issue with Avalonia. (#2651)

Eideren committed:
fix: Ensure teleporting bodies notify bepu's data structure (#2650)

* handle orientation, Add SetPose, and fix center of mass issue with non default rotation

Kryptos-FR committed:
refactor: Nullable and Modernization for Core Presentation projects (#2654)

* [Core.Presentation] Modernize code

* [Core.Design] Make ToEnumerable extension safer

It will not throw a cast exception for a type mismatch, but instead filter out bad items.

* [Core.Quantum] Modernize code

* [Core.Quantum.Tests] Modernize code

* [Core.Presentation.Quantum] Modernize code

* [Core.Presentation.Quantum.Tests] Modernize code

Kryptos-FR committed:
fix: Restore Nuget Resolver UI (#2653)

Kryptos-FR committed:
feat: Run core tests on Linux (#2659)

* [Tests] Allow running tests on Linux

* [Core] Run tests on Linux

* [Core.Presentation] Run tests on Linux

* [Core.Assets] Run tests on Linux

* [BuilEngine] Run tests on Linux

Kryptos-FR committed:
refactor: Nullable and Modernization for Core Assets projects (#2658)

* [Core.BuildEngine.Common] Modernize code

* [Core.Packages] Modernize code

* [Core.Assets.Yaml] Modernize code

* [Build] Remove obsolete Prefer32Bit

* [Core.Assets] Modernize code

* [Core.Assets.Tests] Modernize code

* [Core.Assets.Quantum] Modernize code

* [Core.Assets.Quantum.Tests] Modernize code

* [xunit.runner.stride] Modernize code

* Code review

Jklawreszuk committed:
feat: Enable to build WPF projects outside windows (#2661)

Jklawreszuk committed:
fix: Correct UIPage icon and update default template icon (#2663)

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2668 from stride3d/all-contributors/add-ClamEater14

docs: add ClamEater14 as a contributor for code

allcontributors[bot] committed:
docs: update README.md [skip ci]

allcontributors[bot] committed:
docs: update .all-contributorsrc [skip ci]

VaclavElias committed:
Merge pull request #2669 from stride3d/all-contributors/add-net2cn

docs: add net2cn as a contributor for code

laske185 committed:
fix: add only new asset folders to the package (#2674)

Closes #2667

NicusorN5 committed:
fix: Fix crash when deleting a folder of assets (#2673)

* Fix crash when deleting a folder of assets

VaclavElias committed:
docs: XML Comments added for Bepu Constraints (#2633)

* docs: XML Comments added

* docs: XML comments added for CenterDistanceLimitConstraintComponent.cs

* fix: XML comment wording corrected

* docs: XML comment added for CenterDistanceConstraintComponent

* feat: XML comments added for BallSocketConstraintComponent.cs

* docs: XML comments improvements

* docs: XML comment updated

* fix: Duplicated summary removed

* docs: XML comments improvements

* docs: XML comments improvements

* docs: XML comments improvements

Jklawreszuk committed:
feat: Multiple IDE's support (#2683)

Jklawreszuk committed:
refactor: Modernize script templates (#2662)

azeno committed:
feat: Adds native dependencies for win-arm64 (#2644)

* Adds native dependencies for win-arm64

- The used versions of those dependencies are often rather dated and didn't build out of the box on win-arm64 / Visual Studio 2022, minimal patches had to be applied
- As far as I understood PVRTexLib is delivered as binary only and there's no win-arm64 - therefor added a check to not load it at all. In the end it's only needed by the asset compiler, which in turn is platform neutral. However when invoking it on a arm64 machine, the `dotnet path/to/asset/compiler` will launch it in arm64. If there would be some option to launch it in x64 we wouldn't have any issues with asset compiling. Still the editor might have issues, since its running in process?
- Removed the FreeImageNET load - it's not used anywhere in code
- Fixes some PInvoke signatures in DxtWrapper

* Fixes build of Stride.VisualStudio.sln

Probably introduced by ddc7423145a1d9ac5dd330cb0948240c179611d8

* Removes ARM64 from UWP

* Updates checkout scripts of freetype and Bullet and NativePath

Eideren committed:
fix: Interpolation discontinuities after teleporting a body component (#2681)

johang88 committed:
Vulkan concurrency and validation error fixes (#2684)

* fix: NoSampler is now point instead of linear to prevent validation layer errors.

NoSampler was created using the default linear sampler, this caused validation errors
when using Texture.Load with Texture3d(r16b16f) as it does not support linear sampling.

The sampler was changed to use a point sampler instead which prevents the errors and should cause no issues.

* fix: Fixed Vulkan concurrency issues

Added locking to upload buffer creation and internal queue submit as 0
they are called from multiple threads if texture streaming is eanbled.

* Fixed race condition

Jklawreszuk committed:
fix: Update file location for #2683 (#2688)

Jklawreszuk committed:
Update README.md (#2692) following #2644

Add requirement for ARM

azeno committed:
fix: output directories of DxtWrapper and re-builds with MSVC v143 (#2694)

The dlls were built using `sources\tools\Stride.TextureConverter.Wrappers\build-deps.bat` inside of developer command prompt of VS2022

Eideren committed:
fix: Prevent editor-inserted filters from persisting across views (#2691)

Eideren committed:
feat: User-defined assets support (ScriptableObject like data types and more) (#2527)

Co-authored-by: Nicolas Musset <nicolas@color-rise.xyz>
Co-authored-by: Virgile Bello <virgile@stride3d.net>

Jklawreszuk committed:
fix: Wayland crash workaround (#2698)

Jklawreszuk committed:
refactor: Remove obsolete code (#2666)

Jklawreszuk committed:
refactor: Remove "Refactor" project (#2686)

Eideren committed:
chore: Update bepu to 2.5.0-beta.25 (#2705)

Eideren committed:
fix: Non-default RenderGroup and changing RenderGroup at runtime for UI (#2697)

Eideren committed:
fix: Handle initial collection items appropriately (#2702)

Eideren committed:
feat: Cleanup editor UI for CollidableComponents (#2700)

Eideren committed:
feat: SetTargetPose and deprecate setters for Position and Orientation (#2701)

ourabigdev committed:
docs: Update README.md for ARM64 latest version is called ARM64/ARM64EC (#2715)

Eideren committed:
fix: internal exception being thrown when selecting a EntityComponent type in interface properties + drag and drop support (#2719)

MikhailArsentevTheSecond committed:
docs: Use different logos in README.md for different themes (#2726 & #2728)

Eideren committed:
fix: Avoid race condition in parallelized contact event collection (#2722)

Eideren committed:
fix: AlphaBlendColorValue key retrieved as <Color4> through reflection causing issues when serialized (#2714)

VaclavElias committed:
docs: README.md - Update Visual Studio 2022 installation notes (#2724)

- Added a note that Visual Studio 2022 workloads can be installed via Visual Studio Build Tools.
- Updated C++ build tools section to clarify default enabled status.
- Emphasized that ARM64/ARM64EC build tools are not enabled by default and provided instructions for enabling them.
- Adjusted formatting for improved clarity.

Doprez committed:
feat: Create Bepu demo sample templates. (#2704)

Feralnex committed:
fix: Updated Vortice.Vulkan package to the latest version (#2497)

* Fixed getting actual bounds and refresh rate of the output (monitor/display) from GraphicsOutput.CurrentDisplayMode when using Direct3D

* Updated Vortice.Vulkan package to the latest version

* Use VkUtf8String instead of ReadOnlySpans

* Small refactor with pointers scope in mind

* Added exception logging when failed to create Direct3D device in GraphicsOutput.

* Return null if current format doesn't match the requested one

* CurrentDisplayMode initialization refactor

* Minimize GC collections

* Update version to 2.1.2

* Technical verification remarks

* Specify apiVersion when creating Vulkan instance

Eideren committed:
fix: yaml, rollback trimming of the override tag when parsing collection index to ensure inputs can be re-used (#2718)

Eideren committed:
fix: Optimize assembly reloading (#2717)

tebjan committed:
feat: Enable rendering to HDR displays on DirectX/Windows (#2711)

Eideren committed:
fix: Generate serializers for Non-Private PropertyKey<T> (#2731)

Eideren committed:
fix: Validate Debug.Assert and remove as the logic was mistakenly inverted (#2754)

TranquilAbyss committed:
fix: Added fallback for latent Infinite UIElement size cases (#2765)

hoelzl committed:
fix: compile.bat edgecase when computing duration (#2733) (#2755)

Eideren committed:
docs: Add documentation to bepu constraints (#2763)

Eideren committed:
fix: GameStudio, Prevent locking the mouse cursor when selecting (#2126)

Eideren committed:
fix: Bepu, fix constraint tracking (#2775)

VaclavElias committed:
fix: TransformProcessor.cs - Correct method name typo (#2774)

- Fixed a typo in the method name from `UpdateTransfromationsRecursive()` to `UpdateTransformationsRecursive()` in `TransformProcessor.cs`.

VaclavElias committed:
docs: README.md, add a blank space to match prerequisite name (#2773)

Eideren committed:
fix: Bepu, ctor with valid orientation for constraints (#2777)

VaclavElias committed:
feat: CompoundCollider.cs - Add XML documentation comments (#2778)

Arc-huangjingtong committed:
chore: Rename Xenko -> Stride in .editorconfig (#2780)

Kryptos-FR committed:
feat: Add GitHub CI workflows (#2789)

Arc-huangjingtong committed:
feat: Add Opacity to window api (#2771)

* Update GameWindow.cs

* [WIP]Update Inherit

* Change Opacity value type float->double

Change Opacity value type float->double ，because the internal API use the double type

* Fix comment and Completed the opacity code

* Update XML grammar improvements

* Delete API BackgroundColor

* Delete the useless namespace

* Update GameWindowWinforms.cs

Kryptos-FR committed:
fix: Launcher build (#2783)

Eideren committed:
fix: UI depth bias does not account for facing direction (#2791)

Kryptos-FR committed:
feat: Improve ToString performance for Core.Mathematics structs (#2788)

kutal10 committed:
feat: Bullet Physics, adds on impulses to contact point (#2792)

Kryptos-FR committed:
feat: Improve CI workflows (#2794)

Kryptos-FR committed:
feat: Setup concurrency for CI workflows (#2807)

Kryptos-FR committed:
fix: concurrency group names can collide when running from the CI (main) workflow (#2809)

Eideren committed:
chore: Increment PublicVersion Build number following the breaking change introduced in #2714 (#2822)

laske185 committed:
fix: add null check for oldNode in GameEditor

- Added a null check for `oldNode` before invoking `visitor.Visit(oldNode)`.
- This change prevents potential null reference exceptions during execution.

laske185 committed:
feat: ToggleButton with mouse over images (#2867)

- Updated `ToggleButton.cs` to include new properties for mouse over images:
  - `CheckedMouseOverImage`
  - `UncheckedMouseOverImage`
  - `IndeterminateMouseOverImage`
  - Enhanced `ToggleButtonImageProvider` to consider mouse over images for sprite selection

- Modified `DefaultToggleButtonRenderer.cs` to utilize the new `ToggleButtonImage` property, simplifying the rendering logic

- Expanded `ToggleButtonTests.cs` with new unit tests to cover:
  - Default property values
  - Image handling for different states
  - Mouse over behaviors
  - Introduced `TestableToggleButton` for simulating mouse over states and click events

Eideren committed:
feat: Raycast against collidable (#2861)

Eideren committed:
fix: Bepu and InputSystem update order (#2848)

C0dingSteve committed:
chore: Rename _nativeIntertia to _nativeInertia #2840 (#2847)

Eideren committed:
fix: Bepu, normalize default servo settings across servo constraints (#2838)

Jklawreszuk committed:
feat: Update Stride dev version for the game samples (#2836)

* feat: Update stride dev version for the game samples

* fixup! feat: Update stride dev version for the game samples

Eideren committed:
feat: UI Button color (#2815)

Doprez committed:
feat: Load Services only on first call (#2808)

* make services only load when called.

* added missing proprties

VaclavElias committed:
docs: Add 'behavioral-change' label to breaking changes (#2870)

laske185 committed:
feat: [UI] Add gap support to layout panels (#2855)

azeno committed:
fix: crash when resizing render window and at the same time creating textures in a background thread (#2842)

A relevant lock statement on the internal GraphicsDevice.Resources collection was missing.

MEEMexe committed:
fix: removed double brackets in error log string (#2871)

Arc-huangjingtong committed:
refactor: Discard internal use of FastList (#2798)

VaclavElias committed:
feat: Copilot Instructions (#2857)

* docs: copilot-instructions.md - Add guidelines for Copilot usage

- Added a new section titled "Stride Game Engine Copilot Instructions" to outline best practices for using GitHub Copilot and Copilot Chat.
- Included coding and contribution guidelines to ensure code quality and consistency.
- Provided specific instructions for reviewing pull requests to maintain project standards and team productivity.

* feat: Instructions updated

VaclavElias committed:
docs: Update documentation in BodyComponent and IServo (#2876)

allcontributors[bot] committed:
docs: add laske185 as a contributor for code (#2877)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
docs: add MikhailArsentevTheSecond as a contributor for code (#2878)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
docs: add hoelzl as a contributor for code (#2879)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
docs: add kutal10 as a contributor for code (#2880)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

noa7 committed:
Base Commit for strip hierarchy

. 1. Changes to MeshConvertor and MeshAssetImporter to be able to import sub meshes in model as individual models in project assets

noa7 committed:
UI checkbox to split or not split

Checkbox option for split at time of import

noa7 committed:
Introducing prefab

noa7 committed:
Prefab + naming

With all fixes suggested https://github.com/stride3d/stride/issues/2553#issuecomment-3296960004

noa7 committed:
Fix extra node "RootNode" Parent

noa7 committed:
Material Fix

noa7 committed:
Material + clean up

Eideren commented:
We're back to the previous issue with the materials, we need to only have the materials that are assigned to an object on that object, not every material that was contained in the original scene.

Also, the pivot position for individual models is not retained, they all have their pivot set to the origin instead of wherever it actually is in the file.

noa7 commented:
are you on latest?  it imports fine my side

<img width="1618" height="904" alt="image" src="https://github.com/user-attachments/assets/3e3990cd-39d3-47cc-b5dc-4e08665aab3b" />

Eideren commented:
The vertices are at the right position in world space, but each mesh's pivot point is not right below the model, like it is in the file, it is set to the origin. 
For example, if you *only* place the bookends model in the world, setting its entity's position to 0,0,0 should center the object around the origin
<img width="824" height="623" alt="image" src="https://github.com/user-attachments/assets/4c67948f-7f24-42ea-9947-f44648955f16" />
<img width="774" height="424" alt="image" src="https://github.com/user-attachments/assets/efd7050c-5111-4143-80a2-e568aff6e76f" />
The prefab entity for a given model should have its position moved to match wherever the model is in the imported file, the model asset itself should not have any transformation applied.

noa7 committed:
Material Fix

noa7 committed:
TRS @ nodeinfo -> Entity

noa7 commented:
Makes sense, thanks for pointing out, This is now fixed.  

With this latest-

1. Only the material applied on mesh are visible in inspector 
2. The relative positiong of each item in prefab based of entity data in scene by storing each mesh's TRS data at its nodeinfo object that prefab is constructed with. And each mesh pivot in scene matches to that in source model

<img width="1458" height="604" alt="Capture" src="https://github.com/user-attachments/assets/4d3c4e01-d9a8-4c8e-8864-bbe2066be85e" />
<img width="1901" height="638" alt="Capture1" src="https://github.com/user-attachments/assets/17c78cb7-17a4-4bbf-b83d-94f6ccbee189" />

@Eideren

noa7 committed:
duplicate name fix

Eideren reviewed (commented):
You haven't thoroughly gone through the LLM's work there it seems, there's a fairly large amount of redundant logic, and comments that are clearly targeted at someone interacting with an LLM, not at describing the logic being implemented. 

I'm okay with reviewing *someone's* work, not going over what an LLM spit out, that is your burden to carry - and you're definitely capable of that, I have seen you write better stuff than this.

I haven't gone through the whole thing yet, but do go over your entire PR and clean things up, there's a lot of low hanging fruits I have not marked.

You could also look into using/enabling nullable, there's a bunch of null checks and tests that are redundant or definitely should throw in the null case.

noa7 committed:
Fixing issues raised on PR

https://github.com/stride3d/stride/pull/2898#pullrequestreview-3250065326

noa7 committed:
Clean ResetMaterialsOnPrefabItems and applying TRS prefab items at MeshConvertor Level

noa7 commented:
No string manipulations by storing source mesh name variable in ModelAsset, checked in latest @Eideren

noa7 committed:
cleanup

Eideren commented:
Most of my comments have not been addressed, moving this to draft while you take care of those. Feel free to set it to ready once it is actually ready for review.

noa7 commented:
> Most of my comments have not been addressed, moving this to draft while you take care of those. Feel free to set it to ready once it is actually ready for review.

Can you tell which ones specifically, I gone through all and made  necessary cahnges!

Eideren commented:
Those listed below the following message https://github.com/stride3d/stride/pull/2898#issuecomment-3323404313

noa7 commented:
> Those listed below the following message [#2898 (comment)](https://github.com/stride3d/stride/pull/2898#issuecomment-3323404313)

Its importing fine for me in a folder on latest and have removed strig manipulations?

Eideren commented:
Yes, I got that, but you haven't gone through the review I made below the message I linked here https://github.com/stride3d/stride/pull/2898#issuecomment-3323404313
This stuff:
<img width="966" height="836" alt="image" src="https://github.com/user-attachments/assets/ceceb66b-7aa9-4782-b0c8-9da5fa3b4dbc" />

Eideren commented on sources/editor/Stride.Assets.Presentation/Templates/ModelFromFileTemplateGenerator.cs:
```diff
@@ -154,11 +236,189 @@ protected override IEnumerable<AssetItem> CreateAssets(AssetTemplateGeneratorPar
+                return;
+
+            var underlyingModel=entityInfo.Models.Where(C=>C.MeshName==asset.MeshName).FirstOrDefault();          
+            var nodeContainingMesh=entityInfo.Nodes.Where(c=>c.Name== underlyingModel.NodeName).FirstOrDefault();
```
`Where(x).FirstOrDefault()` can be simplified to `FirstOrDefault(x)`, but if `nodeContainingMesh` or `underlyingModel` are null, the line below would throw, so best use `First(x)`.

noa7 replied:
Fixed

Eideren commented on sources/editor/Stride.Assets.Presentation/Templates/ModelFromFileTemplateGenerator.cs:
```diff
@@ -154,11 +236,189 @@ protected override IEnumerable<AssetItem> CreateAssets(AssetTemplateGeneratorPar
+            if(materialIndices?.Count()< 1)  
+                return; 
+
+            List<ModelMaterial> materialsToApply = null;
+            for (int i = 0; i < asset.Materials.Count; i++)
+            {
+                if (materialIndices.Contains(i))
+                {
+                    (materialsToApply??=new List<ModelMaterial>()).Add(asset.Materials[i]);
+                }
+            }
+
+            asset.Materials.Clear();
+            materialsToApply?.ForEach(_mat => asset.Materials.Add(_mat));          
```
This is more complicated than necessary, do reverse for loop and remove from `asset.Materials` when `materialIndices.Contains(i) == false` instead

noa7 replied:
Fixed

Eideren reviewed (commented):
Well, I did mention `ResetMaterialsOnPrefabItems` requiring a cleanup, it hasn't changed much, I'll leave these few comments for now

noa7 committed:
Addressing PR issues Eideren  #2898

noa7 commented on sources/editor/Stride.Assets.Presentation/Templates/ModelFromFileTemplateGenerator.cs:
```diff
@@ -154,11 +236,195 @@ protected override IEnumerable<AssetItem> CreateAssets(AssetTemplateGeneratorPar
+            return new AssetItem(UPath.Combine(targetLocation, prefabUrl), prefab);
+        }
+
+        private static string SanitizePart(string s)
```
Done!

noa7 commented on sources/editor/Stride.Assets.Presentation/Templates/ModelFromFileTemplateGenerator.cs:
```diff
@@ -154,11 +236,195 @@ protected override IEnumerable<AssetItem> CreateAssets(AssetTemplateGeneratorPar
+            var underlyingModel=entityInfo.Models.Where(C=>C.MeshName==underlyingMeshName).FirstOrDefault();
+            var nodeContainingMesh=entityInfo.Nodes.Where(c=>c.Name== underlyingModel.NodeName).FirstOrDefault();
+
+            var materialIndices=entityInfo.NodeNameToMaterialIndices?.Where(c=>c.Key== nodeContainingMesh.Name)?.FirstOrDefault().Value;
```
Fixed, The FBX as format doesn’t enforce unique mesh names. Adding an extra identifier simply ensures uniqueness. It should work fine, and if there is edge case realted to this not taken into account and might fail, should be easy to take care of!

noa7 committed:
More Fixes #2898

noa7 commented on sources/editor/Stride.Assets.Presentation/Templates/ModelFromFileTemplateGenerator.cs:
```diff
@@ -154,11 +236,195 @@ protected override IEnumerable<AssetItem> CreateAssets(AssetTemplateGeneratorPar
+
+            var materialIndices=entityInfo.NodeNameToMaterialIndices?.Where(c=>c.Key== nodeContainingMesh.Name)?.FirstOrDefault().Value;
+        
+            if(materialIndices?.Count()< 1)  
```
only to account for edge case if theres is no material imported for mesh

Eideren commented on sources/tools/Stride.Importer.3D/MeshConverter.cs:
```diff
@@ -67,6 +75,47 @@ private void ResetConversionData()
+        {
+            keptMeshIndex = -1;
+            keptMeshNameHint = null;
+
+            try
+            {
+                var name = System.IO.Path.GetFileNameWithoutExtension(vfsOutputFilename);
+                if (string.IsNullOrEmpty(name))
+                    return;
+
+                // "(All)" => all meshes
+                if (name.EndsWith(" (All)", StringComparison.OrdinalIgnoreCase))
+                    return;
+
+                // "(Mesh X)" => keep mesh X-1 (legacy)
+                const string tag = " (Mesh ";
+                if (name.EndsWith(")", StringComparison.Ordinal))
+                {
+                    var start = name.LastIndexOf(tag, StringComparison.OrdinalIgnoreCase);
+                    if (start >= 0)
+                    {
+                        var numStr = name.Substring(start + tag.Length, name.Length - (start + tag.Length) - 1);
+                        if (int.TryParse(numStr, out var oneBased) && oneBased >= 1)
+                        {
+                            keptMeshIndex = oneBased - 1;
+                            return;
+                        }
+                    }
+                }
+
+                // Otherwise: remember the full output name; we will try to match a mesh by name later
+                keptMeshNameHint = name;
+            }
+            catch
+            {
+                keptMeshIndex = -1;
+                keptMeshNameHint = null;
+            }
+        }
+
```
I can tell that this is not going to work, far too fragile, and it doesn't, as soon as a user changes the name of a model, the model changes to another or to the whole model instead.

You can add more properties to the `ModelAsset` if needed.

Eideren reviewed (commented):
Can you test your PR more thoroughly, started reviewing again and found another issue on the first thing I checked.

noa7 committed:
Remove Filename based import determination

noa7 commented:
@Eideren can you point which naming issue?

noa7 committed:
Fixing naming issues, storing mesh id as key

noa7 committed:
comm

noa7 committed:
Mesh ModelAsset reimport

noa7 committed:
Fixing multiple material issue

noa7 commented:
Fixed material mapping mismatch between Stride and Assimp.
Stride importer has one material per mesh, while Assimp exports multiple submeshes for multi-material meshes is what makes this tricky. Updated the mesh conversion logic to merge these correctly so materials now display correctly. @Eideren

Eideren commented:
It's visible on your screen, half of the furniture is not on the right spot in the prefab

noa7 committed:
sub mesh material and position fix

Eideren commented:
Looks like it's almost ready for me to review the code, just have to fix two edge case I found:
- Importing a model as *not* split adds a `(All)` suffix to the asset name.
- Importing a model, then assigning a new material to the file, saving it, going back to the game studio and doing right click > `Update Selected Assets From Source`, does not create a new entry under the list of materials in the model asset.

You can use this new file to test the second issue: [Untitled.zip](https://github.com/user-attachments/files/23290808/Untitled.zip)
Import the version you already have, make a backup of that file, then overwrite it with the one above and `Update Selected Assets From Source`. `Untitled-Object_0_1` should go from one to five material entries.

noa7 committed:
Fix all

noa7 commented:
> Looks like it's almost ready for me to review the code, just have to fix two edge case I found:
> 
> * Importing a model as _not_ split adds a `(All)` suffix to the asset name.
> * Importing a model, then assigning a new material to the file, saving it, going back to the game studio and doing right click > `Update Selected Assets From Source`, does not create a new entry under the list of materials in the model asset.
> 
> You can use this new file to test the second issue: [Untitled.zip](https://github.com/user-attachments/files/23290808/Untitled.zip) Import the version you already have, make a backup of that file, then overwrite it with the one above and `Update Selected Assets From Source`. `Untitled-Object_0_1` should go from one to five material entries.

- Removed the suffix "All".

The problem with second one is the update  and creation pipe line are different, even for the existing pipeline(seperate from split heirarchy) on updating asset from source will not add the extra materials, cause the class responsible for creating asset ModelFromFileTemplateGenerator is not involved in update, it will update the model itself but will not create any additional asset.
Currently updating the asset from source is updating asset itself. May be for now we can keep it as it and have another project to unify update and create, as this could it might be a significant change to the logic flow. Please advice. @Eideren

noa7 committed:
Material slots sync upon update from source

noa7 commented:
Checked in update to best accommodate this, it should show empty slots based on asset but I am not sure how much it would hold as updating material this way can lead to potential problems. The structure of imported file after changing materials is different in terms of how mesh and fbx nodes are connected to sync back with data original file in terms of matching which mesh and material to current comes with all sorts of problems are it has to account all ways it can changing, naming, order, hierarchy or any other way user can tinker file in 3d model tool. Updating entire model (split hierarchy) from source is fine as this matching to what exiting is not needed!

@Eideren
